Navigation: Integrate search into topnav

ashharrison90 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

- integrates search into topnav
  - this basically reuses the existing search page and surfaces it in a modal.
  - modal is opened on click of the search bar/when typing into the search bar (if you've tabbed into the search bar for example)
- refactors `searchQueryReducer` to be hooked up to the redux store and use `createSlice`
  - hooking up to the store is necessary so that the 2 inputs can share the state
- creates a new component, `DashboardSearchModal`, which is basically a copy of the current `DashboardSearch`
  - this gives us a bit more flexibility to make changes only relevant to topnav
  - means we can delete the `DashboardSearch` component when we deliver topnav
  
# behaviour

https://user-images.githubusercontent.com/20999846/189173776-5557f87e-6bac-4013-a3e9-4eb4a1bc4528.mp4

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes https://github.com/grafana/grafana/issues/54837

**Special notes for your reviewer**:

ashharrison90 committed:
behaviour mostly there

ashharrison90 committed:
slight performance improvement

ashharrison90 committed:
slightly nicer...

ashharrison90 committed:
refactor search and add it to the store

ashharrison90 committed:
add comments about removing old component

ashharrison90 committed:
remove unneeded logic

torkelo committed:
small design tweak

torkelo committed:
More small tweaks

torkelo committed:
Restore top margin

torkelo commented:
Nice! I think this works really well.

* The border-radius is a bit extreme (compared to anything else in Grafana where we usually stay around 2px - 4px) 
* On a really wide monitor it look a bit odd with the input stil showing behind it (easy fix is probably to hide original input)
* Feel like closing search should unfocus the navbar search input and not show text there / clear action. Think the top search input only needs to be a trigger that opens the search. Having to close search and hit clear action and click away to unfoucs the input feels like tree actions that should all be just one action. 

![Screenshot from 2022-09-08 18-39-16](https://user-images.githubusercontent.com/10999/189179259-f73c64cc-478e-4b8c-982f-790a9ee85b9a.png)

Pushed a small design tweak
* Change to primary background (think this works better, the topnav background is primary). Think we can get rid fo the padding around the result container but leaving that for future design change. 
* Reduced border radius a bit.

ashharrison90 committed:
add onCloseSearch/onSelectSearchItem to useSearchQuery

ashharrison90 commented:
@torkelo tweaks look good, thanks! 🙌 

> On a really wide monitor it look a bit odd with the input stil showing behind it (easy fix is probably to hide original input)

i've just removed the max-width on the modal for now so it'll take 70% of the full width of the screen. helps with making the animation consistent at least. happy to iterate on this as the modal probably looks a little spaced out now. 

> Feel like closing search should unfocus the navbar search input and not show text there / clear action. Think the top search input only needs to be a trigger that opens the search. Having to close search and hit clear action and click away to unfoucs the input feels like tree actions that should all be just one action.

have made some changes here which i think(?) is the best of both worlds:
- removed restoring the focus to the topnav input when closing the search modal
- cleared the search input **when something is selected from the search modal**
- if closing the search modal by clicking the x/clicking outside/pressing escape, the value is not cleared

this feels most natural to me, and mirrors something like slack's search behaviour.

side note: i've noticed a fairly annoying bug when directly opening a url with the search open (e.g. http://localhost:3000/?orgId=1&search=open&query=foo). the animation probably shouldn't run at all in this case, but it runs twice and is kinda jarring. i'll work on a fix, but i don't think it needs to hold out this PR 🙂

ashharrison90 committed:
Merge branch 'main' into ash/54837

torkelo reviewed (approved):
Sounds good!
